_posts/2025-01-22-code-repair.md

```diff
@@ -146,35 +70,118 @@ index 9af4721c5e32..b7d45890c453 100644
+The Cray-LM framework fine-tunes a pre-trained Llama model using a dataset formatted as JSON lines. Each sample includes fields such as:
+
+`['bug_report_path', 'bug_report_text', 'diff_path', 'diff_text', 'source_code_path', 'line_number', 'code']`
+
+This input is preprocessed and a few lines above and below the offending line_number are used as additional context. This information is then fed to the train the model.
+
+```python
+import masint
+import jsonlines
+
+# Retrieve source code snippet with line numbers
+def get_source_code(data, before_lines=5, after_lines=5):
+    lines = data["code"].split("\n")
+    start_line = max(0, data["line_number"] - before_lines)
+    end_line = min(len(lines), data["line_number"] + after_lines)
+    return "\n".join(lines[start_line:end_line])
+
+# Prepare training data
+def get_data(training_data_file, dataset_size=1000):
+    with jsonlines.open(training_data_file) as reader:
+        raw_data = list(reader)
+
+    return [
+        {
+            "input": prompt_template.format(
+                source_code_path=entry["source_code_path"],
+                line_number=entry["line_number"],
+                code=get_source_code(entry),
+                bug_report_text=entry["bug_report_text"]
+            ),
+            "output": entry["diff_text"]
+        }
+        for entry in raw_data[:dataset_size]
+    ]
+
+# Main function
+def main():
+    data = get_data(training_data_file=args.input)
+    llm = masint.SupermassiveIntelligence()
+    llm.train(data, train_args={"max_steps": 20, "learning_rate": 3e-3})
+```
 
+### 4. Automated Fix Generation
+The fine-tuned LLM generates code patches based on new bug reports.
+```python
+import jsonlines
+import masint
+
+def load_data(eval_file_path):
+    with jsonlines.open(eval_file_path) as reader:
+        return list(reader)
+
+def get_source_code(data, before_lines=5, after_lines=5):
+    lines = data["code"].split("\n")
+    start_line = max(0, data["line_number"] - before_lines)
+    end_line = min(len(lines), data["line_number"] + after_lines)
+    return "\n".join(lines[start_line:end_line])
+
+def get_dataset(data):
+    return [
+        prompt_template.format(
+            source_code_path=entry["source_code_path"],
+            line_number=entry["line_number"],
+            code=get_source_code(entry),
+            bug_report_text=entry["bug_report_text"]
+        )
+        for entry in data
+    ]
+
+def main():
+    data = load_data(eval_file_path=args.input)
+    dataset = get_dataset(data)
+
+    llm = masint.SupermassiveIntelligence()
+
+    for entry in dataset:
+        results = llm.generate(prompts=[entry])
+        print(f"\n{results}")
+```
+### 5. Validation
+Generated fixes are manually compared against reference diffs to validate accuracy and alignment with coding standards. This enhanced workflow demonstrates how LLMs can be leveraged to automate bug detection and repair, reducing both development time and security risks.
```

Comment:
@greg1232 I have not added the specifics on {llama model, Lora, accuracy numbers}. Can you help me with this section to add it appropriately>